File: apps/web/src/commands/timeline/element/shift-split-remainder.ts
import { Command, type CommandResult } from "@/commands/base-command";
import type { MediaTime } from "@/wasm";
import { MoveElementCommand } from "./move-elements";
import type { SplitElementsCommand } from "./split-elements";

/**
 * Moves the right-side elements produced by a SplitElementsCommand to a new
 * start time. Must run after the split inside a BatchCommand: the right-side
 * element ids only exist once the split has executed. Rebuilds the inner move
 * on every execute() so redo picks up the fresh ids a re-executed split
 * generates.
 */
export class ShiftSplitRemainderCommand extends Command {
	private inner: MoveElementCommand | null = null;

	constructor({
		split,
		newStartTime,
	}: {
		split: SplitElementsCommand;
		newStartTime: MediaTime;
	}) {
		super();
		this.split = split;
		this.newStartTime = newStartTime;
	}

	private readonly split: SplitElementsCommand;
	private readonly newStartTime: MediaTime;

	execute(): CommandResult | undefined {
		const rightSideElements = this.split.getRightSideElements();
		if (rightSideElements.length === 0) {
			return undefined;
		}
		this.inner = new MoveElementCommand({
			moves: rightSideElements.map((element) => ({
				sourceTrackId: element.trackId,
				targetTrackId: element.trackId,
				elementId: element.elementId,
				newStartTime: this.newStartTime,
			})),
		});
		return this.inner.execute();
	}

	undo(): void {
		this.inner?.undo();
	}
}

File: rust/crates/commands/src/timeline/element/shift_split_remainder.rs
use std::cell::RefCell;
use std::rc::Rc;

use scene::{ElementRef, Project};
use time::MediaTime;

use crate::command::{Command, CommandError, CommandResult};
use crate::timeline::element::move_elements::{MoveElementCommand, PlannedElementMove};

pub struct ShiftSplitRemainderCommand {
    right_side: Rc<RefCell<Vec<ElementRef>>>,
    new_start_time: MediaTime,
    inner: Option<MoveElementCommand>,
}

impl ShiftSplitRemainderCommand {
    pub fn new(right_side: Rc<RefCell<Vec<ElementRef>>>, new_start_time: MediaTime) -> Self {
        Self {
            right_side,
            new_start_time,
            inner: None,
        }
    }
}

impl Command for ShiftSplitRemainderCommand {
    fn execute(
        &mut self,
        project: &mut Project,
    ) -> Result<Option<CommandResult>, CommandError> {
        let right_side = self.right_side.borrow().clone();
        if right_side.is_empty() {
            return Ok(None);
        }

        let mut inner = MoveElementCommand::new(
            right_side
                .iter()
                .map(|element| PlannedElementMove {
                    source_track_id: element.track_id.clone(),
                    target_track_id: element.track_id.clone(),
                    element_id: element.element_id.clone(),
                    new_start_time: self.new_start_time,
                })
                .collect(),
            Vec::new(),
        );
        let result = inner.execute(project)?;
        self.inner = Some(inner);
        Ok(result)
    }

    fn undo(&mut self, project: &mut Project) {
        if let Some(inner) = &mut self.inner {
            inner.undo(project);
        }
    }
}
